virtcontainers/netns_watcher/netns_watcher.c

```diff
@@ -0,0 +1,1213 @@
+		return -1;
+	}
+
+	if ((ret = ioctl(sock, SIOCGIFHWADDR, &ifr)) == -1) {
+		printf("Failed to get HW ADDR: %s", strerror(errno));
+		goto exit;
+	}
+
+	nif->hw_addr = hw_addr_to_string((unsigned char *)
+					ifr.ifr_hwaddr.sa_data);
+
+	if ((ret = ioctl(sock, SIOCGIFMTU, &ifr)) == -1) {
+		printf("Failed to get MTU: %s", strerror(errno));
+		goto exit;
+	}
+
+	if (asprintf(&(nif->mtu), "%d", ifr.ifr_mtu) == -1) {
+		printf("Failed allocating memory: %s\n",
+		       strerror(errno));
+		goto exit;
+	}
+
+exit:
+	close(sock);
+	return ret;
+}
+
+/*
+ * Fill a new interface from the information given by ifaddrs.
+ */
+int add_ip_addr_to_iface_list(struct ifaddrs *ifa, unsigned int idx)
+{
+	unsigned char family;
+	int s;
+	char host[NI_MAXHOST];
+
+	family = ifa->ifa_addr->sa_family;
+	if (family != AF_INET && family != AF_INET6) {
+		return 0;
+	}
+
+	s = getnameinfo(ifa->ifa_addr,
+			(family == AF_INET) ? sizeof(struct sockaddr_in) :
+					      sizeof(struct sockaddr_in6),
+			host, NI_MAXHOST, NULL, 0, NI_NUMERICHOST);
+	if (s) {
+		printf("Failed getnameinfo(): %s\n", gai_strerror(s));
+		return -1;
+	}
+
+	insert_ip_addr(idx, host, family);
+
+	return 0;
+}
+
+/* 
+ * Scan the network and store it into a global list.
+ *
+ * A precise description of the network is needed to make sure events received
+ * through the netlink socket will be properly interpreted.
+ */
+int scan_netns()
+{
+	struct ifaddrs *ifaddr, *ifa;
+	int n;
+	int ret = 0;
+
+	if (getifaddrs(&ifaddr) == -1) {
+		printf("Failed getifaddrs(): %s\n", strerror(errno));
+		return -1;
+	}
+
+	for (ifa = ifaddr, n = 0; ifa != NULL; ifa = ifa->ifa_next, n++) {
+		struct iface nif;
+		unsigned int if_idx;
+
+		if (!ifa->ifa_name) {
+			continue;
+		}
+
+		if_idx = if_nametoindex(ifa->ifa_name);
+		if (!if_idx) {
+			printf("Failed if_nametoindex() for %s: %s\n",
+			       ifa->ifa_name, strerror(errno));
+			goto exit;
+		}
+
+		if (iface_list[if_idx].idx == INIT_IF_IDX) {
+			nif.name = strdup(ifa->ifa_name);
+
+			ret = create_iface_from_ifaddrs(ifa, &nif);
+			if (ret) {
+				free_iface(&nif);
+				goto exit;
+			}
+
+			nif.idx = if_idx;
+
+			if (if_idx >= MAX_IFACES) {
+				printf("Interface idx %d over the limit %d\n",
+				       if_idx, MAX_IFACES);
+				free_iface(&nif);
+				goto exit;
+			}
+
+			/* Update the global list of interfaces */
+			iface_list[if_idx] = nif;
+		}
+
+		ret = add_ip_addr_to_iface_list(ifa, if_idx);
+		if (ret) {
+			goto exit;
+		}
+	}
+
+exit:
+	freeifaddrs(ifaddr);
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int ret = 0;
+
+	struct watcher_params params = {
+		.netns_path   = NULL,
+		.pod_id       = NULL,
+		.runtime_path = NULL,
+	};
+
+	/* Initialize list of interfaces */
+	init_iface_list();
+
+	/* Validate parameters */
+	ret = valid_params(argc, (const char**) argv, &params);
+	if (ret) {
+		return ret;
+	}
+
+	/* Enter network namespace */
+	ret = enter_netns((const char*) params.netns_path);
```

Comment:
I think we can run the monitor via `network.run(networkNSPath string, cb func() error) error`  instead of `enter_netns((const char*) params.netns_path)`.

Reply:
You mean that virtcontainers would start the process directly into the network namespace, and I agree this would work, but I think it is better to have this being performed from the C process directly as entering namespaces is safer in C than Go. See https://github.com/kata-containers/runtime/issues/148 for more details about namespaces issues in Go.

Moreover, this makes this program more generic as it is able to enter a namespace on its own.